### ISSUE TITLE
Request support for Powerworld PW58330 heat pump

### ISSUE THREAD

**Issue by dslava70** [make-all/tuya-local#1530]
# Log Message

I have created a config file for a heat pump device pw58330 from the PowerWorld company.

# Information about DPS mappings
Config file (yaml) attached as txt file: 
[powerworld_pw58330_heatpump.txt](https://github.com/make-all/tuya-local/files/13952108/powerworld_pw58330_heatpump.txt)

# Product ID
Model: Low Gwp R290 Monoblock Heat Pumps
Website:  https://www.powerworldhp.com/heat-pump/home-heat-pump/low-gwp-r290-monoblock-heat-pumps.html

# Information about how the device functions
R290 full DC monoblock heat pump
[China Custom Low Gwp R290 Monoblock Heat Pumps Manufacturers Suppliers.pdf](https://github.com/make-all/tuya-local/files/13949750/China.Custom.Low.Gwp.R290.Monoblock.Heat.Pumps.Manufacturers.Suppliers.pdf)

**Comment by Andrea-IT** [comment]
Hi, I also have a PowerWorld heat pump, I would like to know if you have integrated it into home assistant, tuya intergation find the cloud device, but do not show the entity :-(
Thanks see you soon

**Comment by nankoboerma** [comment]
I've got the same issue with my PowerWorld PW58410 heatpump.
Working fine in the Smartlife app, but adding it to HA via the Tuya, LocalTuya or Tuya Local app gives me nothing but a simple switch.

Did you guys find a solution?

**Comment by make-all** [comment]
There really isn't enough information here to make a good quality integration. The submitted config is just a collection of switches, select entities and sensors, with no climate or water_heater control. It is missing temperature adjustments completely.

There is no further information available to improve on this, so it would help if you can gather logs and iot.tuya.com info from the QueryThingsDataModel API call, if you are interested in getting support for this.

**Comment by nankoboerma** [comment]
Thanks for your reply!
Here is some info on the heatpump  model:
   "model": "PW58410",
      "name": "My house heat pump",
      "product_id": "74jzzoqr2ghxyu1p",
      "product_name": "My house heat pump",

This is the information from the QueryThingsDataModel API:
{
  "result": {
    "model": "{\"modelId\":\"fc1fls\",\"services\":[{\"actions\":[],\"code\":\"\",\"description\":\"\",\"events\":[],\"name\":\"默认服务\",\"properties\":[{\"abilityId\":1,\"accessMode\":\"rw\",\"code\":\"switch\",\"description\":\"\",\"extensions\":{\"iconName\":\"icon-dp_power2\",\"attribute\":\"1664\"},\"name\":\"开关\",\"typeSpec\":{\"type\":\"bool\"}},{\"abilityId\":2,\"accessMode\":\"rw\",\"code\":\"mode\",\"description\":\"智能；强劲；静音\",\"extensions\":{\"iconName\":\"icon-dp_mode\",\"attribute\":\"1152\"},\"name\":\"频率模式\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"smart\",\"strong\",\"mute\"]}},{\"abilityId\":5,\"accessMode\":\"rw\",\"code\":\"work_mode\",\"description\":\"热水模式；制热模式；制冷模式；热水+制热；热水+制冷\",\"extensions\":{\"iconName\":\"icon-dp_mode\",\"attribute\":\"1152\"},\"name\":\"运行模式\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"wth\",\"heat\",\"cool\",\"wth_heat\",\"wth_cool\"]}},{\"abilityId\":6,\"accessMode\":\"rw\",\"code\":\"temp_unit_convert\",\"description\":\"摄氏度；华氏度\",\"extensions\":{\"iconName\":\"icon-dp_mode\",\"attribute\":\"1152\"},\"name\":\"温标切换\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"c\",\"f\"]}},{\"abilityId\":15,\"accessMode\":\"ro\",\"code\":\"fault\",\"description\":\"\",\"extensions\":{\"iconName\":\"icon-baojing\",\"scope\":\"fault\",\"attribute\":\"1152\"},\"name\":\"故障告警\",\"typeSpec\":{\"type\":\"bitmap\",\"label\":[\"Er03\",\"Er04\",\"Er05\",\"Er06\",\"Er09\",\"Er10\",\"Er12\",\"Er14\",\"Er15\",\"Er16\",\"Er18\",\"Er20\",\"Er21\",\"Er23\",\"Er26\",\"Er27\",\"Er29\",\"Er32\",\"Er33\",\"Er34\",\"Er42\",\"Er62\",\"Er63\",\"Er64\",\"Er66\",\"Er67\",\"Er68\",\"Er69\",\"Er70\",\"Er73\"],\"maxlen\":30}},{\"abilityId\":101,\"accessMode\":\"ro\",\"code\":\"status_parameter_group_1\",\"description\":\"\",\"name\":\"状态参数组1\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":110,\"accessMode\":\"rw\",\"code\":\"hot_water_set\",\"description\":\"\",\"name\":\"热水设定温度\",\"typeSpec\":{\"type\":\"value\",\"max\":99,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":111,\"accessMode\":\"rw\",\"code\":\"heating_setting\",\"description\":\"\",\"name\":\"制热设定温度\",\"typeSpec\":{\"type\":\"value\",\"max\":99,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":112,\"accessMode\":\"rw\",\"code\":\"cooling_setting\",\"description\":\"\",\"name\":\"制冷设定温度\",\"typeSpec\":{\"type\":\"value\",\"max\":99,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":118,\"accessMode\":\"rw\",\"code\":\"parameter_group_1\",\"description\":\"此DP点包含20个参数，1个参数占4个字节（参考数值型）\",\"name\":\"参数组1\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":119,\"accessMode\":\"rw\",\"code\":\"parameter_group_2\",\"description\":\"\",\"name\":\"参数组2\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":120,\"accessMode\":\"rw\",\"code\":\"parameter_group_3\",\"description\":\"\",\"name\":\"参数组3\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":121,\"accessMode\":\"rw\",\"code\":\"parameter_group_4\",\"description\":\"\",\"name\":\"参数组4\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":122,\"accessMode\":\"rw\",\"code\":\"parameter_group_5\",\"description\":\"\",\"name\":\"参数组5\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":123,\"accessMode\":\"rw\",\"code\":\"parameter_group_6\",\"description\":\"\",\"name\":\"参数组6\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":124,\"accessMode\":\"rw\",\"code\":\"parameter_group_7\",\"description\":\"\",\"name\":\"参数组7\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":125,\"accessMode\":\"wr\",\"code\":\"reset\",\"description\":\"按键操作在关机下有效且下发完成后复位该功能\",\"name\":\"一键复位默认值\",\"typeSpec\":{\"type\":\"bool\"}},{\"abilityId\":126,\"accessMode\":\"rw\",\"code\":\"parameter_group_8\",\"description\":\"\",\"name\":\"参数组8\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":140,\"accessMode\":\"rw\",\"code\":\"parameter_group_23\",\"description\":\"电量统计专用参数组标识符必须使用 \\nparameter_group_23\",\"name\":\"电量统计专用参数组\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":199,\"accessMode\":\"ro\",\"code\":\"custom_fault_bit\",\"description\":\"\",\"name\":\"自定义故障\",\"typeSpec\":{\"type\":\"value\",\"max\":999999999,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"\"}}]}]}"
  },
  "success": true,
  "t": 1728805475266,
  "tid": "fdd24501893611efbc016e7818a0f8a2"
}

I got some entities working (on localtuya, with manual DP), but not the parameter_groups. Some deliver RAW data in HA, but i cant figure out how to translate it.

I hope there can be a way to import the heatpump into HA. 
Thanks in advance for your help!

Kind regards N

**Comment by make-all** [comment]
There are still no logs, so a high probability that the config will not match, but a first pass has been done. Note in the QueryThingsData API, there is no temperature sensor dps, only the temperature setting dps. This is the opposite from the submitted config, which had more temperature sensors than was reasonable, but no setting.

I don't know if the submitted config was just copy and paste of irrelevant sensors, or it is incompatible with the PW58410, but for now I've gone with the PW58410 for the installed config, as QueryThingsDataModel info is generally accurate.

**Comment by dslava70** [comment]
In my homeassistant (core) the config I provided above has been working just fine.... 
You can setup it in your homeassistant manually like this:
1. Rename powerworld_pw58330_heatpump.txt to powerworld_pw58330_heatpump.yaml
2. Copy it to /home/homeassistant/.homeassistant/custom_components/tuya_local/devices as homeassistant user
3. Restart Homeassistant

**Comment by dslava70** [comment]
You should be able to see a new integration among the tuya_local devices in the Integrations of Homeassistant

**Comment by dslava70** [comment]
On the https://eu.platform.tuya.com/ my device is visible as:
Product Name 复制-PW58330
Product Category znrb

**Comment by dslava70** [comment]
I am not sure if that is important but also on the https://eu.platform.tuya.com/ I made this change:
1. On the Configure Control Instruction Mode page https://platform.tuya.com/cloud/setMode?id=....  
I have changed "Standard Instruction" setting to "DP Instruction"

![image](https://github.com/user-attachments/assets/b82173a5-c5ea-4927-b58c-26e7d9001141)

**Comment by dslava70** [comment]
Thus, in my Homeassistant I can see the integration with my PW heatpump as follows:
![image](https://github.com/user-attachments/assets/763de0ae-7fcd-44cc-92fa-deb01e1354a8)

**Comment by nankoboerma** [comment]
Hi, thanks for all your reply's.
I added the file Powerworld_pw58330_heatpump as you said. (also make a similar file named Powerworld_pw58410_heatpump)
But when i add the device again in tuya local i don't get a good device type in the menu:

![image](https://github.com/user-attachments/assets/66b0b578-1ad3-4f3c-9720-d9b5fc324d5e)

**Comment by nankoboerma** [comment]
This is my log, i hope it is something you can work with. I'm not familiar with it..
[log.txt](https://github.com/user-attachments/files/17355656/log.txt)

**Comment by make-all** [comment]
Thank you, the added config appears to match that log.

```
powerworld_pw58410_waterheater matched 100%
  water_heater:
   operation_mode: Hot water + heating
   work_mode: wth_heat
   temperature_unit: C
   temperature: 50
   parameter_group_1: None
   parameter_group_2: None
   parameter_group_3: None
   parameter_group_4: None
   parameter_group_5: None
   parameter_group_6: None
   parameter_group_7: None
   parameter_group_8: None
   power_statistics: None
  climate_heater:
    preset_mode: comfort
    hvac_mode: heat
    temperature_unit: C
    temperature: 30
    cool_temp: 12
  select_temperature_unit:
    option: celsius
  binary_sensor_problem:
    sensor: False
    fault_code: 0
    custom_fault_bit: 0
  button_factory_reset:
    button: None
```

**Comment by make-all** [comment]
The screenshots for the PW58330 device seem to show that intemp and outtemp are instructions (settable values), so these may be related to the water and heating temperatures.

It would be useful, since you appear to have access to iot.tuya.com to provide the QueryThingsDataModel result and logs from this integration as requested in the new device issue template

**Comment by dslava70** [comment]
yes, my model is this:

{
  "result": {
    "model": "{\"modelId\":\"e1ept4\",\"services\":[{\"actions\":[],\"code\":\"\",\"description\":\"\",\"events\":[],\"name\":\"默认服务\",\"properties\":[{\"abilityId\":1,\"accessMode\":\"rw\",\"code\":\"switch\",\"description\":\"\",\"extensions\":{\"iconName\":\"icon-dp_power2\",\"attribute\":\"1664\"},\"name\":\"开关\",\"typeSpec\":{\"type\":\"bool\"}},{\"abilityId\":2,\"accessMode\":\"rw\",\"code\":\"mode\",\"description\":\"智能；强劲；静音\",\"extensions\":{\"iconName\":\"icon-dp_mode\",\"attribute\":\"1152\"},\"name\":\"运行模式\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"smart\",\"strong\",\"mute\"]}},{\"abilityId\":5,\"accessMode\":\"rw\",\"code\":\"work_mode\",\"description\":\"热水模式；制热模式；制冷模式；热水+制热；热水+制冷\",\"extensions\":{\"iconName\":\"icon-dp_mode\",\"attribute\":\"1152\"},\"name\":\"工作模式\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"wth\",\"heat\",\"cool\",\"wth_heat\",\"wth_cool\"]}},{\"abilityId\":6,\"accessMode\":\"rw\",\"code\":\"temp_unit_convert\",\"description\":\"摄氏度；华氏度\",\"extensions\":{\"iconName\":\"icon-dp_mode\",\"attribute\":\"1152\"},\"name\":\"温标切换\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"c\",\"f\"]}},{\"abilityId\":15,\"accessMode\":\"ro\",\"code\":\"fault\",\"description\":\"\",\"extensions\":{\"iconName\":\"icon-baojing\",\"scope\":\"fault\",\"attribute\":\"1152\"},\"name\":\"故障告警\",\"typeSpec\":{\"type\":\"bitmap\",\"label\":[\"Er03\",\"Er04\",\"Er05\",\"Er06\",\"Er09\",\"Er10\",\"Er12\",\"Er14\",\"Er15\",\"Er16\",\"Er18\",\"Er20\",\"Er21\",\"Er23\",\"Er26\",\"Er27\",\"Er29\",\"Er32\",\"Er33\",\"Er34\",\"Er42\",\"Er62\",\"Er63\",\"Er64\",\"Er66\",\"Er67\",\"Er68\",\"Er69\",\"Er70\"],\"maxlen\":29}},{\"abilityId\":101,\"accessMode\":\"ro\",\"code\":\"intemp\",\"description\":\"\",\"name\":\"进水温度\",\"typeSpec\":{\"type\":\"value\",\"max\":220,\"min\":-50,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":102,\"accessMode\":\"ro\",\"code\":\"outtemp\",\"description\":\"\",\"name\":\"出水温度\",\"typeSpec\":{\"type\":\"value\",\"max\":220,\"min\":-50,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":103,\"accessMode\":\"ro\",\"code\":\"whjtemp\",\"description\":\"\",\"name\":\"环境温度\",\"typeSpec\":{\"type\":\"value\",\"max\":220,\"min\":-50,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":104,\"accessMode\":\"ro\",\"code\":\"cmptemp\",\"description\":\"\",\"name\":\"排气温度\",\"typeSpec\":{\"type\":\"value\",\"max\":400,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":105,\"accessMode\":\"ro\",\"code\":\"hqtemp\",\"description\":\"\",\"name\":\"回气温度\",\"typeSpec\":{\"type\":\"value\",\"max\":220,\"min\":-50,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":106,\"accessMode\":\"ro\",\"code\":\"ptemp\",\"description\":\"\",\"name\":\"外盘管温度\",\"typeSpec\":{\"type\":\"value\",\"max\":220,\"min\":-50,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":107,\"accessMode\":\"ro\",\"code\":\"cptemp\",\"description\":\"\",\"name\":\"制冷盘管温度\",\"typeSpec\":{\"type\":\"value\",\"max\":220,\"min\":-50,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":108,\"accessMode\":\"ro\",\"code\":\"wttemp\",\"description\":\"\",\"name\":\"水箱温度\",\"typeSpec\":{\"type\":\"value\",\"max\":220,\"min\":-50,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":109,\"accessMode\":\"ro\",\"code\":\"step_run\",\"description\":\"\",\"name\":\"主阀膨胀开度\",\"typeSpec\":{\"type\":\"value\",\"max\":500,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"P\"}},{\"abilityId\":111,\"accessMode\":\"ro\",\"code\":\"stepb_run\",\"description\":\"\",\"name\":\"增焓膨胀阀开度\",\"typeSpec\":{\"type\":\"value\",\"max\":500,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"P\"}},{\"abilityId\":112,\"accessMode\":\"ro\",\"code\":\"cmp_cur\",\"description\":\"\",\"name\":\"压机电流\",\"typeSpec\":{\"type\":\"value\",\"max\":65,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"A\"}},{\"abilityId\":113,\"accessMode\":\"ro\",\"code\":\"sink_temp\",\"description\":\"\",\"name\":\"散热片温度\",\"typeSpec\":{\"type\":\"value\",\"max\":220,\"min\":-99,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":114,\"accessMode\":\"ro\",\"code\":\"dc_bus_voltage\",\"description\":\"\",\"name\":\"直流母线电压值\",\"typeSpec\":{\"type\":\"value\",\"max\":380,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"V\"}},{\"abilityId\":115,\"accessMode\":\"ro\",\"code\":\"cmp_act_frep\",\"description\":\"\",\"name\":\"压机实际频率\",\"typeSpec\":{\"type\":\"value\",\"max\":500,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"Hz\"}},{\"abilityId\":116,\"accessMode\":\"ro\",\"code\":\"dc_fan_speed\",\"description\":\"\",\"name\":\"直流风机1风速\",\"typeSpec\":{\"type\":\"value\",\"max\":2000,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"\"}},{\"abilityId\":117,\"accessMode\":\"ro\",\"code\":\"dc_fan2_speed\",\"description\":\"\",\"name\":\"直流风机2风速\",\"typeSpec\":{\"type\":\"value\",\"max\":2000,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"\"}},{\"abilityId\":118,\"accessMode\":\"rw\",\"code\":\"parameter_group_1\",\"description\":\"此DP点包含20个参数，1个参数占4个字节（参考数值型）\",\"name\":\"参数组1\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":119,\"accessMode\":\"rw\",\"code\":\"parameter_group_2\",\"description\":\"\",\"name\":\"参数组2\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":120,\"accessMode\":\"rw\",\"code\":\"parameter_group_3\",\"description\":\"\",\"name\":\"参数组3\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":121,\"accessMode\":\"rw\",\"code\":\"parameter_group_4\",\"description\":\"\",\"name\":\"参数组4\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":122,\"accessMode\":\"rw\",\"code\":\"parameter_group_5\",\"description\":\"\",\"name\":\"参数组5\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":123,\"accessMode\":\"rw\",\"code\":\"parameter_group_6\",\"description\":\"\",\"name\":\"参数组6\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":124,\"accessMode\":\"rw\",\"code\":\"parameter_group_7\",\"description\":\"\",\"name\":\"参数组7\",\"typeSpec\":{\"type\":\"raw\",\"maxlen\":128}},{\"abilityId\":125,\"accessMode\":\"wr\",\"code\":\"reset\",\"description\":\"按键操作在关机下有效且下发完成后复位该功能\",\"name\":\"一键复位默认值\",\"typeSpec\":{\"type\":\"bool\"}}]}]}"
  },
  "success": true,
  "t": 1728826197652,
  "tid": "3d5d07cd896711efb77736033862a1ba"
}

**Comment by nankoboerma** [comment]
> Thank you, the added config appears to match that log.
> 
> ```
> powerworld_pw58410_waterheater matched 100%
>   water_heater:
>    operation_mode: Hot water + heating
>    work_mode: wth_heat
>    temperature_unit: C
>    temperature: 50
>    parameter_group_1: None
>    parameter_group_2: None
>    parameter_group_3: None
>    parameter_group_4: None
>    parameter_group_5: None
>    parameter_group_6: None
>    parameter_group_7: None
>    parameter_group_8: None
>    power_statistics: None
>   climate_heater:
>     preset_mode: comfort
>     hvac_mode: heat
>     temperature_unit: C
>     temperature: 30
>     cool_temp: 12
>   select_temperature_unit:
>     option: celsius
>   binary_sensor_problem:
>     sensor: False
>     fault_code: 0
>     custom_fault_bit: 0
>   button_factory_reset:
>     button: None
> ```

Thanks! Where should is use this config?

**Comment by dslava70** [comment]
I have many tuya devices so it is a mess in my log file related to messages from my tuya devices.... I have tried to filter out ones for my PowerWorld heatpump device.

2024-10-13 16:41:49.122 DEBUG (SyncWorker_4) [custom_components.tuya_local.helpers.device_config] Loaded device config powerworld_pw58330_heatpump.yaml
2024-10-13 16:43:26.937 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] option values: ['Smart', 'Powerful', 'Quiet']
2024-10-13 16:43:26.937 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] PowerWorld PW58330 Heatpump: Mapped dps 2 value from mute to Quiet
2024-10-13 16:43:26.937 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] PowerWorld PW58330 Heatpump: Mapped dps 2 value from mute to Quiet
2024-10-13 16:43:26.937 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] option values: ['Smart', 'Powerful', 'Quiet']
2024-10-13 16:43:26.938 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] option values: ['HotWater', 'Heat', 'Cool', 'HotWater+Heat', 'HotWater+Cool']
2024-10-13 16:43:26.938 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] PowerWorld PW58330 Heatpump: Mapped dps 5 value from heat to Heat
2024-10-13 16:43:26.938 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] PowerWorld PW58330 Heatpump: Mapped dps 5 value from heat to Heat
2024-10-13 16:43:26.938 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] option values: ['HotWater', 'Heat', 'Cool', 'HotWater+Heat', 'HotWater+Cool']
2024-10-13 16:43:26.938 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 101 (sensor), unable to determine valid values
2024-10-13 16:43:26.939 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 102 (sensor), unable to determine valid values
2024-10-13 16:43:26.939 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 103 (sensor), unable to determine valid values
2024-10-13 16:43:26.939 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 104 (sensor), unable to determine valid values
2024-10-13 16:43:26.940 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 105 (sensor), unable to determine valid values
2024-10-13 16:43:26.940 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 106 (sensor), unable to determine valid values
2024-10-13 16:43:26.940 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 107 (sensor), unable to determine valid values
2024-10-13 16:43:26.941 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 108 (sensor), unable to determine valid values
2024-10-13 16:43:26.941 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 109 (sensor), unable to determine valid values
2024-10-13 16:43:26.941 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 111 (sensor), unable to determine valid values
2024-10-13 16:43:26.941 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 112 (sensor), unable to determine valid values
2024-10-13 16:43:26.942 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 113 (sensor), unable to determine valid values
2024-10-13 16:43:26.942 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 114 (sensor), unable to determine valid values
2024-10-13 16:43:26.943 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 115 (sensor), unable to determine valid values
2024-10-13 16:43:26.944 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 116 (sensor), unable to determine valid values
2024-10-13 16:43:26.944 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 117 (sensor), unable to determine valid values
2024-10-13 16:43:26.944 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 118 (sensor), unable to determine valid values
2024-10-13 16:43:26.944 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 118 (sensor), unable to determine valid values
2024-10-13 16:43:26.945 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 119 (sensor), unable to determine valid values
2024-10-13 16:43:26.945 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 119 (sensor), unable to determine valid values
2024-10-13 16:43:26.945 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 120 (sensor), unable to determine valid values
2024-10-13 16:43:26.945 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 120 (sensor), unable to determine valid values
2024-10-13 16:43:26.945 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 121 (sensor), unable to determine valid values
2024-10-13 16:43:26.945 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 121 (sensor), unable to determine valid values
2024-10-13 16:43:26.946 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 122 (sensor), unable to determine valid values
2024-10-13 16:43:26.946 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 122 (sensor), unable to determine valid values
2024-10-13 16:43:26.946 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 123 (sensor), unable to determine valid values
2024-10-13 16:43:26.946 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 123 (sensor), unable to determine valid values
2024-10-13 16:43:26.947 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 124 (sensor), unable to determine valid values
2024-10-13 16:43:26.947 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] No mapping for dpid 124 (sensor), unable to determine valid values
2024-10-13 16:43:27.791 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of value
2024-10-13 16:43:27.791 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of value
2024-10-13 16:43:27.791 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for step of value

**Comment by nankoboerma** [comment]
At this moment i cant figure out what to do?
I've added the JSON files, but it didnt work.
Where should i use the config code of Jason Rumney?

Kind regards

**Comment by make-all** [comment]
Download the file at https://github.com/make-all/tuya-local/blob/main/custom_components/tuya_local/devices/powerworld_pw58410_waterheater.yaml to your `config/custom_components/tuya_local/devices/` directory.

**Comment by nankoboerma** [comment]
Great, its integrated! Thanks!
I got some of the entities visible.
Is there a way how i can add the other parameter(groups)?
![IMG_5834](https://github.com/user-attachments/assets/32a77efb-52f5-4a5b-ad9b-5a8737b055f5)

**Comment by make-all** [comment]
Parameter groups were added as attributes under the water_heater, since they do not seem useful to surface in the UI as sensors.  If you want them in the UI, the tile and entity cards give the option of displaying an attribute instead of the main state of the entity.

**Comment by nankoboerma** [comment]
Hello Jason,

Thanks again!

Could you make it possible to read all the parametergroups etc. as sensors. So that it can be an input value for an automation in the future. I'm hoping to track the power-, temperature- and COP value.
For the most of the parameter groups the value's of the sensor would be RAW date, something like AAAAGwAAACEAAAAKAAA....etc.

Is there a way that we can translate these values to something usefull.

Thanks!!

**Comment by make-all** [comment]
Attributes can also be read by automations.  I don't see any advantage to having them as sensors given the unreadable encoded binary data they contain.

**Comment by dslava70** [comment]
I am using pyscript to decode such values AAAAGwAAACEAAAAKAAA....etc.
This is an encoded string of values... You need to figure out which value in the string belongs to your heatpump sensor value. You can do it by matching the decoded value to a value shown in your Tuya app for the device.
Here is my simple code to decode the param1 and set its values to a sensor(s) from Homeassistant.


This is a content of my write_pw_heatpump_parameter1.py file below...

import base64

@service()
def write_pw_heatpump_parameter1():
    task.unique("write_pw_heatpump_parameter1")
    param1_sensor = "sensor.pw_heatpump_parameter_group_1"  

    try:
        prev_valueBase64 = str(state.get(param1_sensor))
        #log.warning("prev_valueBase64: " + prev_valueBase64)        

        if len(prev_valueBase64) == 108: 
            prev_valueHex = base64.b64decode(prev_valueBase64).hex()
            #log.warning("prev_valueHex: " + prev_valueHex)        
            if len(prev_valueHex) == 160:
                map_Values = map(''.join, zip(*[iter(prev_valueHex)]*8))
                list_Values = list(map_Values)
                if len(list_Values) == 20:
                    #log.warning("list_Values: " + str(list_Values))        
                    log.warning("list_Values[0]: " + str(list_Values[0]) + ' ' + str(to_s16(list_Values[0])) )        
                    log.warning("list_Values[1]: " + str(list_Values[1]) + ' ' + str(to_s16(list_Values[1])) ) 
                    log.warning("list_Values[15]: " + str(list_Values[15]) + ' ' + str(to_s16(list_Values[15])) )   
                    log.warning("list_Values[0] hex: " + str(list_Values[0]) + ' ' + from_s16(to_s16(list_Values[0])) )        
                    log.warning("list_Values[1] hex: " + str(list_Values[1]) + ' ' + from_s16(to_s16(list_Values[1])) ) 
                    log.warning("list_Values[15] hex: " + str(list_Values[15]) + ' ' + from_s16(to_s16(list_Values[15])) )   
                    # state.set("sensor.pw_unit_status", value=prev_valueBase64)
    except Exception as ex:
        log.warning(ex)

def to_s16(value):
    val = int(value, 16)
    return -(val & 0x8000) | (val & 0x7fff)

def from_s16(value):
    return "{:08x}".format((int(value) & 0xFFFFFFFF), '08x')

**Comment by make-all** [comment]
According to the documentation they are all lists of 4 byte values, but there is no description of which values are where.
If some can be figured out, it would be possible to add sensors, or numbers if they are adjustable, for those values using base64 type and a mask for the value.

**Comment by nankoboerma** [comment]
For parameter group 1 i got the following from Query Things Data Model
"This DP point contains 20 parameters, 1 parameter occupies 4 bytes (refer to numerical type)"

**Comment by nankoboerma** [comment]
> r the water_heater, since they do not seem useful to surface in the UI as sensors. If you want them in the UI, the tile an
Sorry DSLAVA70 i am not experienced enough to know what to do with this information, can you maybe write a small instruction?
Thanks in advance!

**Comment by nankoboerma** [comment]
I found a way to translate the first parametergroup 1

![image](https://github.com/user-attachments/assets/2c395b01-ef98-431b-982f-d4cfc608a3aa)

16: retour water temperature
18: output water temperature
14: outside temperature (surrounding)
21: exhaust gas temperature
24: retour gas temperature
10: temperature evaporator
12: temperature cooling battery
49: temperature WTH boiler
350: pressure (P) expansionvalve
0: pressure (P) assistant expansionvalve
0: power (A) compressor
40: temp cooling body
566: DC bus (Volts)
0: compression frequentie (Hz)
0: Windrotationspeed 1 (rpm)
0: Windrotationspeed 2 (rpm)
5: pressuresensor (Bar)
10: temperature low pressure
0: PV signal
0: SG signal

**Comment by nankoboerma** [comment]
DP 140 (parametergroup 23) shows the following:
![image](https://github.com/user-attachments/assets/77709553-5352-42b6-bf4e-64d510516ec1)

0: Heating capacity (KW)
0: Water Flowrate (m³/h)
0: power debit (A)
231: voltage (V)
0: power usage (W)
0: COP (coeficient of performance, very important!)
0: set percentage of rotational speed DC-waterpump (%)
0: actual rotional speed DC-waterpump (%)

**Comment by nankoboerma** [comment]
Could you guys please, please, please give me a way on which i can get the values in the above parameter groups 1 and 23 into separate sensors (normal numers) in home assistant. This way i can log the performance of the heatpump, which is not possible in the native tuya app.

I hope you've got the needed info from my translation from above.

Thanks in advance!

**Comment by nankoboerma** [comment]
Thanks Jason!
I've got most of the sensors visible.
I added the log for you to check.
![New sensors](https://github.com/user-attachments/assets/ea36bd49-bc05-4f9f-9a0c-1af66299daf1)


[tuya_local-01JA60Y74P85BMAHM1X7Q9TXMX-Water heat pump N 141024-925a46d73f891ff2cc9d293274f80f40.json](https://github.com/user-attachments/files/17396707/tuya_local-01JA60Y74P85BMAHM1X7Q9TXMX-Water.heat.pump.N.141024-925a46d73f891ff2cc9d293274f80f40.json)
[log 241016.txt](https://github.com/user-attachments/files/17396708/log.241016.txt)

**Comment by nankoboerma** [comment]
![image](https://github.com/user-attachments/assets/f37163a9-07ad-4bb8-9274-383ad80723ed)
Turned the heatpump on and several values started working.
Heating capacity and Volume flow need an ajustment on the decimal point

**Comment by nankoboerma** [comment]
Incredible job!! I'm so happy with your help.
All the sensors show the information now!
I'll keep tracking the coming days and share the result.

Thank you so much!

![image](https://github.com/user-attachments/assets/4fc529ba-8c48-498a-912a-939ac6bf28d1)

**Comment by nankoboerma** [comment]
Works great for the last couple of days. I noticed that the cop has to be adjusted with a decimal; from for example 50 > 5.0
![IMG_5854](https://github.com/user-attachments/assets/f65e7063-7306-4c00-a9a8-e8968fee0d75)

**Comment by DaasMichiel** [comment]
Hi, I'm new here I also have a power world PW58410 heat pump which is only installed as a heating hvac system. 

However, when using tuya-local I am not able to turn on the heat pump using the climate.water_heat_pump_heater. 
But the water_heater.water_heat_pump is able to turn the system on when it is set to heating. 

I am completely new to home assistant, so I have attempted to change the code in the YAML files to attempt to imitate device turn on and turn off from the water heater in the climate files, but I have had no progress yet. Could you maybe help me in fixing this? :)

<img width="490" alt="image" src="https://github.com/user-attachments/assets/ca9493ed-d275-4bc2-94b7-f166bd00d1e6">

**Comment by make-all** [comment]
I have added an additional config for using the PW58410 device primarily for HVAC only.
The same can be done for PW58330 and other similar devices that are already supported if the demand is there.

**Comment by dslava70** [comment]
Yes, please add a config for the PW58410 device primarily for HVAC only.
Also, I have added missing properties which can be retrieved from the base64 parameters. Names of the properties are taken from the Tuya app for the device. 
Additionally, I have found an issue in the tuya_local code - after applying a mask for the the base64 value, if the value is a negative  value (signed int), tuya_local displays in an unsigned int resulting that -20 is dysplayed as 4294967281.  I have temporarily applied a local fix to in my tuya_local version until it is fixed in tuya_local.

[powerworld_pw58330_waterheater.txt](https://github.com/user-attachments/files/17506275/powerworld_pw58330_waterheater.txt)

**Comment by DaasMichiel** [comment]
> I have added an additional config for using the PW58410 device primarily for HVAC only. The same can be done for PW58330 and other similar devices that are already supported if the demand is there.

Thank you so much! It works perfectly now :)

**Comment by nankoboerma** [comment]
With updating tuya local in HA, the config text for the device PW58410 changed. It misses the sensor masking in the Base 64 parametergroups. 
I manually copied the text into it again.. and the different sensors are back.

**Comment by netforceatg** [comment]
Just want to say thanks to everyone for adding support for this. Works very well with home assistant

Works well with my Powerworld R290 11kW Air Source to Water heat pump. (PW040-DKZLRS-E/S)

However I have discovered a bug. I use the pump in silent mode(use the operator panel/HMI to set Silent mode). When I do this the tuya local integration shifts to 'Sleep' mode on the unit.

I assume this is due to the mapping from the thermostat card in home assistant is mapped from Sleep to Silent.

When the thermostat card in home assistant for the unit is in sleep mode and I change temperature, the temperature changes. But after a few seconds it always goes back to 35 degrees. (even tho the display and Tuya app shows the new temperature)

Any way to fix this?

**Comment by nankoboerma** [comment]
Yeah, the ‘sleep’ is synonymous for silent in the JSON file of the API i believe. So its just the name i guess.
Regarding the temperature, i dont use it so cant help you with that. Kind regards

**Comment by make-all** [comment]
It seems the new temperature is not fed back from the device. The integration will show the local value for a few seconds or until the device sends an update, but if no update is received, it will assume the device did not accept the new value.

This could be because the receive connection has been blocked by another local protocol user, or it may be a problem with the device.

**Comment by ctcDNightmare** [comment]
> Yes, please add a config for the PW58330 device for HVAC too. Also, I have added missing properties which can be retrieved from the base64 parameters. Names of the properties are taken from the Tuya app for the device. Additionally, I have found an issue in the tuya_local code - after applying a mask for the base64 value, if the resulting value is a negative value (signed int), tuya_local displays it as an unsigned int resulting that -15 is displayed as 4294967281. I have temporarily applied a local fix in my tuya_local version until it is fixed in tuya_local.
> 
> [powerworld_pw58330_waterheater.txt](https://github.com/user-attachments/files/17506275/powerworld_pw58330_waterheater.txt)

I've got a similar model (PW58386) and combining bits n pieces from your config as well as the PW58410 I was able to get all the basics working BUT: Does your model report those base64/parameter_group_x DPs or can you send base64 values back to set the target temperature for example? 

```yaml
  - entity: sensor
    name: Heating target Temperature
    class: temperature
    category: diagnostic
    dps:
      - id: 118
        type: base64
        optional: true
        name: sensor
        scale: 0.1
        mapping:
          - mask: "00000000000000000000000000000000FFFFFFFF00000000\
              000000000000000000000000000000000000000000000000\
              000000000000000000000000000000000000000000000000\
              0000000000000000"
```
As a sensor, it will only get populated if I change any setting on the pump itself or via app, but apart from that it never show up, not even in the cloud/device logs

I tried moving thise value into the hvac part as 'target_temperature' and as soon as I use the app to update ANY parameter and force a report of the "parameter_group_X" DPs I see the value on the climate card, but changing the dial does nothing on the device itself :'(

I played around with tinytuya and UPDATEDPS or DP_QUERY to no success.
Normal logs when nothing is happening via app or physically:
![image](https://github.com/user-attachments/assets/cde5cc3a-2dd0-48cb-a5df-de4e7237c7b7)

Logs just when I change a setting in the app after dp140 appeared in the logs:
![image](https://github.com/user-attachments/assets/f589e759-cb20-4dd9-aa8e-452d5e1ad733)

My current progress of figuring out what is what with those parameter_groups
[notes.md](https://github.com/user-attachments/files/17741913/notes.md)

My working havc-only config so far
[powerworld_pw58386_hvaconly.txt](https://github.com/user-attachments/files/17741933/powerworld_pw58386_hvaconly.txt)

And just for another datapoint: export from tuya query properties:
[tuya-query-properties.txt](https://github.com/user-attachments/files/17742015/tuya-query-properties.txt)


Any ideas on the matter?

**Comment by mbulles** [comment]
Hi, I have a Power World PW060 heatpump and I'm looking to achieve the same as some of you already achieved, which means getting the right data in HA. Almost all of the data [DP] looks similar so that should not be the issue.

Unfortunately I'm missing some pieces of information to get it working.

I have retrieved my device local key via tinytuya.

When trying to add the device by using local tuya the device us automatically recognised and the key is automatically added as well. And I have configured 1 DP [ the on/off button].

What is not clear to me yet, is how to get the rest of the heatpump data/entities in HA/Local tuya?
I have tried to install the YAML file mentioned in the given location [ do I need to change the device ID in the yaml file?]
The file location can not be found in my HA setup, the location I have is 
/config/custom_components/localtuya there is no devices directory.

If maybe can provide a little more step by step approach or hints to get it to work that would be awesome!

Thanks,

**Comment by mbulles** [comment]
Hi, I have a Power World PW060 heatpump and I'm looking to achieve the same as some of you already achieved, which means getting the right data in HA. Almost all of the data [DP] looks similar so that should not be the issue.

Unfortunately I'm missing some pieces of information to get it working.

I have retrieved my device local key via tinytuya.

When trying to add the device by using local tuya the device us automatically recognised and the key is automatically added as well. And I have configured 1 DP [ the on/off button].

What is not clear to me yet, is how to get the rest of the heatpump data/entities in HA/Local tuya without manually adding all the missing entities?
I have tried to install the YAML file mentioned in the given location [ do I need to change the device ID in the yaml file?]
The file location can not be found in my HA setup, the location I have is
/config/custom_components/localtuya there is no devices directory.

If maybe someone can provide a little more step by step approach or hints to get it to work that would be awesome!

Thanks,

**Comment by ctcDNightmare** [comment]
> Hi, I have a Power World PW060 heatpump and I'm looking to achieve the same as some of you already achieved, which means getting the right data in HA. Almost all of the data [DPS] looks similar so that should not be the issue.
> 
> Unfortunately I'm missing some pieces of information to get it working.
> 
> I have retrieved my device local key via tinytuya.
> 
> When trying to add the device by using local tuya the device us automatically recognised and the key is automatically added as well. And I have configured 1 DP [ the on/off button].
> 
> What is not clear to me yet, is how to get the rest of the heatpump data/entities in HA/Local tuya without manually adding all the missing entities? I have tried to install the YAML file mentioned in the given location [ do I need to change the device ID in the yaml file?] The file location can not be found in my HA setup, the location I have is /config/custom_components/localtuya there is no devices directory.
> 
> If maybe someone can provide a little more step by step approach or hints to get it to work that would be awesome!
> 
> Thanks,

Common trap I also stumbled into:

local-tuya and tuya-local are two different things :)
The one where you only get 1 DP for on/off is local-tuya and over here with the custom yaml files is tuya-local which works with the above config and more DPs

**Comment by mbulles** [comment]
Thanks for the quick response;

But when looking in HACS I can only find these two;

[image: image.png]

Can you maybe provide me a link where I can find add tuya local, I looked
but it seems I'm always being sent to LocalTuya integration.

Thanks,

**Comment by mbulles** [comment]
Hi again, 

Thanks for quick reply!
I have installed the tuya-local via HACS and I was able to install the heatpump using the pw58830 profile + waterheater.

There are however 3 values I noticed which do not seem to correct in the Diagnose overview;

1. Air return temperature,
2. Evaporator Coil and
3. Low Pressure

![heatpump](https://github.com/user-attachments/assets/ad708fba-6e0b-46f1-8486-7dfc845e63be)

Is there a way I can fix this, or does this need a fix on tuya-local end?
Thanks for all the help!!!!

**Comment by DaasMichiel** [comment]
Hey, I noticed now that temperatures are getting lower that some of the temperature sensor values seem to be translated by home assistant to unsigned 32bit integers. Is it possible to change this such that the sensor values are interpreted as signed integers?

**Comment by make-all** [comment]
No, this is not generally possible when values are encoded together in a blob like this device does.

**Comment by ctcDNightmare** [comment]
> No, this is not generally possible when values are encoded together in a blob like this device does.

well cound't we, in theory,  have a flag to tell toya-local parse that it's a signed?

like using an S to indicate the sign bit and after extracting the bits, move the indicated bit << all the way left to have a signed 32bit afterwards?
```yaml
        mapping:
          - mask: "0000000000000000000000000000000000000000SFFFFFFF\
              000000000000000000000000000000000000000000000000\
              000000000000000000000000000000000000000000000000\
              0000000000000000"
```
or having signed: true besides type: base64?

**Comment by ctcDNightmare** [comment]
That's the exact problem I have / was talking about in this thread. It will never get updated by the pump on it's own, only when we send it a new value (for the whole group), change it via app or physically at the pump. 
Happens with all parameter_groups_1 to _8

**Comment by DaasMichiel** [comment]
> It will never get updated

"it" being the parameter groups 1 to 8? because for me those do update frequently, for me it is the "Heating_setting_temperature" that takes long to update. But parameter_group_1, where this value interestingly enough is also stored, updates immediately upon changing the set temperature. I believe that the smart home app that is normally used with these pumps retrieves its information from this parameter group rather than the "heating_setting_temperature" because of this delay.

**Comment by ctcDNightmare** [comment]
Well what are we talking about when talking about "delay" ? minutes? hours? days?
For me, it just won't send it's param_1 to _8 if I just leave stuff alone...
![local-network-normal-operations](https://github.com/user-attachments/assets/4884aee6-f303-479e-81b6-fdc205da61bc)

Only when I touch any controls in app/pump/via scripts will it get updated:
![local-network-after-changing-any-user-parameter](https://github.com/user-attachments/assets/34d7f63e-200b-4adf-a45b-e9f6d1761b12)

And the app gets the requests from the cloud if you're out of your wifi or switches to local wifi when at home.

**Comment by Allu95** [comment]
Hi, I am new to home assistant and especially with the Tuya Local. I was using Local Tuya and it was working okay. The problem is that I cannot get the 

> Hi, thanks for all your reply's. I added the file Powerworld_pw58330_heatpump as you said. (also make a similar file named Powerworld_pw58410_heatpump) But when i add the device again in tuya local i don't get a good device type in the menu:
> 
> ![image](https://private-user-images.githubusercontent.com/93254624/376035844-66b0b578-1ad3-4f3c-9720-d9b5fc324d5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIzOTQwNjgsIm5iZiI6MTczMjM5Mzc2OCwicGF0aCI6Ii85MzI1NDYyNC8zNzYwMzU4NDQtNjZiMGI1NzgtMWFkMy00ZjNjLTk3MjAtZDliNWZjMzI0ZDVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIzVDIwMjkyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFiY2Y5ZTc5MTgwMjZiN2Q0MDBkMDdhMTQ4OTNhYmZlMDYxZTM2ZWQxNTg2ZmQxMjUyYWNhMDMyZDdmYWVjZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tOEdpYmqY0jzl4Dz7QoLh3Ospowq_4zvWgNAOI2k3jg)

Hi, how you fixed this problem? I have been struggling for hours because I cannot get the device under the device type drop down menu.

**Comment by crazy-ahmet** [comment]
Hello. I also have a poolworld heatpump. I´ve installed homeassistant and linked with tuya cloud (not local tuya and not tuya local). Is it possible to get this running with this since the heatpump is not in my local network?

The device shows up in home assistant as model "my house heat pump (unsupported).
I don´t get any datapoints from it. I´ve also iobroker which gets the data every minute from the cloud. here it works but i want all data incl. the groups data which need some more decoding, therefore tried with homeassistant but have not much knowledge with this system.

**Comment by make-all** [comment]
If there is a route to the network the heatpump is on, then it will work, but you will need to manually specify the IP address, as it will not be able to be discovered automatically.

**Comment by crazy-ahmet** [comment]
I wonder why it doesn´t work pulling the data from the clould since iobroker does this every minute.

i can establish dynamic ip adress and portforwarding if i know the ports. how to install this config file?

**Comment by make-all** [comment]
This is a local integration, it does not pull data from the cloud.
If you have an issue with this integration, I suggest you open a new issue if you can describe it clearly enough with enough detail that someone without access to your device can debug it.
Otherwise if you have questions, or cannot describe enough details about the issue to expect a solution, discussions would be a better venue.

This issue is already closed.

**Comment by alesica** [comment]
I got a PW58329 about two years ago. Powerworld has modbus documentation describing all registers / groups. I managed to read all important registers via local tuya, but that was the easy part. The main problem was to write some important parameters like the target temperature or the compressor frequency, as they come in groups of 8 in BASE64. To be honest other than the mode (heat vs domestic hot water), the temperature and occasionally the compressor frequency there isn't much else that needs to be configured. I have it running on weather compensation and I set the min compressor frequency based on the ambient temp, as by default is set to 40Hz but it could go down to 30Hz which works really for mild temperatures (ambient temp >9C) drawing a mere 800-900W

**Comment by nankoboerma** [comment]
@make-all are there new breaking changes that caused the following issue?
Since some days the connection with the Powerworld Heatpump is lost and i get the following error when reconnecting;
<img width="702" alt="Image" src="https://github.com/user-attachments/assets/fee4ef7e-37b0-4a4a-a8a6-a708555d5316" />

I've checked the Tuya Developer API and the local key and IP adres are still correct.

I'm looking forward to your advise,
Kind Regards N

**Comment by mbulles** [comment]
Hi,

Fyi, my powerworld does still work via Localtuya.

Op do 5 jun 2025, 11:40 schreef nankoboerma ***@***.***>:

**Comment by nankoboerma** [comment]
> [@make-all](https://github.com/make-all) are there new breaking changes that caused the following issue? Since some days the connection with the Powerworld Heatpump is lost and i get the following error when reconnecting; <img alt="Image" width="702" src="https://private-user-images.githubusercontent.com/93254624/451810593-fee4ef7e-37b0-4a4a-a8a6-a708555d5316.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAzMTg3MDMsIm5iZiI6MTc1MDMxODQwMywicGF0aCI6Ii85MzI1NDYyNC80NTE4MTA1OTMtZmVlNGVmN2UtMzdiMC00YTRhLWE4YTYtYTcwODU1NWQ1MzE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjE5VDA3MzMyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1YzliMWFhZTU0NGNlMWM2ZDQyOTFkYzY4NGMzODhjZjA2NjVkNDA3OTkwZDRjZjA1NmNjOWEwZWI1ZmYxZWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uAOmqbag1Icr64dlMFTxxBB1ah6WOhxJ3XtGunH3JmQ">
> 
> I've checked the Tuya Developer API and the local key and IP adres are still correct.
> 
> I'm looking forward to your advise, Kind Regards N

It is fixed, probably a local network problem.

**Comment by DimitrijL** [comment]
> I got a PW58329 about two years ago. Powerworld has modbus documentation describing all registers / groups. I managed to read all important registers via local tuya, but that was the easy part. The main problem was to write some important parameters like the target temperature or the compressor frequency, as they come in groups of 8 in BASE64. To be honest other than the mode (heat vs domestic hot water), the temperature and occasionally the compressor frequency there isn't much else that needs to be configured. I have it running on weather compensation and I set the min compressor frequency based on the ambient temp, as by default is set to 40Hz but it could go down to 30Hz which works really for mild temperatures (ambient temp >9C) drawing a mere 800-900W

@alesica , can you provide the Modbus documentation? I would also be interested in knowing how to set the minimum frequency of the compressor. I would also like to be able to clock at 30 Hz...

**Comment by Dark-Noir** [comment]
> I got a PW58329 about two years ago. Powerworld has modbus documentation describing all registers / groups. I managed to read all important registers via local tuya, but that was the easy part. The main problem was to write some important parameters like the target temperature or the compressor frequency, as they come in groups of 8 in BASE64. To be honest other than the mode (heat vs domestic hot water), the temperature and occasionally the compressor frequency there isn't much else that needs to be configured. I have it running on weather compensation and I set the min compressor frequency based on the ambient temp, as by default is set to 40Hz but it could go down to 30Hz which works really for mild temperatures (ambient temp >9C) drawing a mere 800-900W

Can you share the powerworld modbus documentation and how you get that all working in HA?
I consider such a heat pump atm., but only if i can control it.

**Comment by netforceatg** [comment]
> > I got a PW58329 about two years ago. Powerworld has modbus documentation describing all registers / groups. I managed to read all important registers via local tuya, but that was the easy part. The main problem was to write some important parameters like the target temperature or the compressor frequency, as they come in groups of 8 in BASE64. To be honest other than the mode (heat vs domestic hot water), the temperature and occasionally the compressor frequency there isn't much else that needs to be configured. I have it running on weather compensation and I set the min compressor frequency based on the ambient temp, as by default is set to 40Hz but it could go down to 30Hz which works really for mild temperatures (ambient temp >9C) drawing a mere 800-900W
> 
> Can you share the powerworld modbus documentation and how you get that all working in HA? I consider such a heat pump atm., but only if i can control it.

@alesica I'm also looking to control the PowerWorld Heatpump over modbus, so any info about your project would be helpful

Would be great to either add a WiFi/Ethernet ModbusTCP to RTU(RS485) to the PowerWorld heatpump and an addon to homeassistant to control it, or alternativly have a esp with modbusRTU chip loaded with esphome or similar acting as a bridge between home assistant and the PowerWolrd PW58329
